src/poetry/puzzle/provider.py

```diff
@@ -947,6 +940,22 @@ def debug(self, message: str, depth: int = 0) -> None:
 
             self._io.write(debug_info)
 
+    def _group_by_source(
+        self, dependencies: Iterable[Dependency]
+    ) -> list[list[Dependency]]:
+        groups: list[list[Dependency]] = []
+        for dep in dependencies:
+            for group in groups:
+                if (
+                    dep.is_same_source_as(group[0])
+                    and dep.source_name == group[0].source_name
```

Comment:
I find it somewhat surprising that `is_same_source_as` doesn't compare `source_name` -- however, I expect that that is for a good reason?

In general, that method could use a good docstring and more detailed comments -- however, that is a discussion for a different day (and repo).

Reply:
At first, I found it surprising, too. However, after further thinking it makes sense because `is_same_source_as` is a method of `PackageSpecification` and `source_name` is an attribute of Dependency which is a subclass of `PackageSpecification`. We could override `is_same_source_as` for `Dependency` to consider `source_name` if a Dependency is compared to another `Dependency` and not consider it if a `Dependency` is compared to a `Package` or `PackageSpecification`.

Reply:
Ah, that makes sense -- I think this change would be easier to reason about if we got that into core first.

Reply:
python-poetry/poetry-core#488

Reply:
It seems like we can't change that easily. See failing downstream tests in the core PR. Sometimes `source_name` is not relevant because dependencies created from packages via `to_dependency()` can't have a source name and the solver still has to consider such dependencies equal to dependencies with source name.

Reply:
Looks like a battle for another day, then. Could you update your core PR to instead explain that gotcha in a comment somewhere for future reference?